Jetpack Social: Create Jetpack Social no connection dashboard card

wargcm opened the pull request:
See: #20974 

## Description

Adds a new dashboard card, `DashboardJetpackSocialCardCell`. It currently only shows the no connections social view but will later be updated to include the view when the user is out of shares.

## Screenshots

| Light | Dark |
|------|------|
| ![Screenshot 2023-07-24 at 2 57 39 PM](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/a0c3a252-b4ac-4ca0-8578-813ad8e83435) | ![Screenshot 2023-07-24 at 2 57 51 PM](https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/da39d878-f4ed-48a5-99f4-bb07ca5076f8) |

## Testing

### No connections dashboard card
- Launch Jetpack and login if needed
- Select a blog that supports social connections and also has none setup
- Navigate to the home dashboard
- Scroll down if necessary
- **Verify** the no connections dashboard card appears with UI matching Th1ahHKq53k5JT1PNDMavY-fi-833%3A13918
- **Verify** the dashboard card's position is below the stats card and above drafts
- Tap on `Connect your profiles`
- Setup a connection and then navigate back to the dashboard
- **Verify** the card no longer appears
- Remove the connection that was just setup (Menu > Social)
- Navigate back to the home dashboard
- Scroll down if necessary
- On the no connections dashboard card, tap on the context menu `...`
- Tap on `Hide this`
- **Verify** the card is now hidden
- Switch to a different blog with no social connections
- **Verify** the card appears for this blog
- Switch back to the original blog
- **Verify** the card is still hidden

### Post pre publishing & post settings quick check

- Select a blog that supports social connections and also has none setup
- Create a new blog post
- Navigate to the post settings screen (`...` > `Post Settings`)
- **Verify** the no connections view still appears the same as before
- Navigate back to the new blog post
- Type something to enable the `Publish` button
- Tap on `Publish`
- In the pre publishing sheet, **verify** the no connections view still appears the same as before

## Regression Notes
1. Potential unintended areas of impact
The no connections view on the pre publishing sheet & post settings screen.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
- Tests for the display logic of the card
- Test for the display state of the card

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

wargcm committed:
Create Jetpack Social no connection dashboard card

wargcm commented on WordPress/Classes/ViewRelated/Jetpack/Social/JetpackSocialNoConnectionView.swift:
```diff
@@ -33,6 +34,7 @@ struct JetpackSocialNoConnectionView: View {
                 }
             }
         }
+        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
```
This was a fun one to figure out 🙃. It fixes the following issues:

1. The `UIHostingController`'s view had extra space causing issues with the insets. `maxWidth` & `maxHeight` here expand the SwiftUI view to the entire host's view.
2. Even with the view fully expanded now, it was being aligned to the center of the view. This also caused issues with the insets. Changing the alignment to `.topLeading` makes the insets work as you'd expect.

dvdchr replied:
Yeah, there's a lot of quirky behaviors for sure. I've also stumbled on this weird behavior and ended up seeing [this SO post](https://stackoverflow.com/questions/69265914/on-ios-15-the-uihostingcontroller-is-adding-some-weird-extra-padding-to-its-hos) (there's a comment from Salim, too! 🕵🏼 ) which mentions that it's fixed on iOS 16. But I guess you're testing on an iOS 16 Simulator? 🙃 

Anyways, glad you've found a workaround 👍🏼

wargcm replied:
> But I guess you're testing on an iOS 16 Simulator?

Yep, my simulators are `16.4` so I'm not sure if it was fixed with iOS 16 or if this was a different issue.

dvdchr commented on WordPress/WordPressTest/Blog Dashboard/Cards/DashboardJetpackSocialCardCellTests.swift:
```diff
@@ -0,0 +1,154 @@
+
+    private let featureFlags = FeatureFlagOverrideStore()
+
+    override func setUp() {
+        super.setUp()
+
+        try? featureFlags.override(FeatureFlag.jetpackSocial, withValue: true)
+    }
+
+    override func tearDown() {
+        super.tearDown()
+
+        try? featureFlags.override(FeatureFlag.jetpackSocial, withValue: FeatureFlag.jetpackSocial.originalValue)
+    }
```
Nice! 👍🏼

dvdchr commented on WordPress/WordPressTest/Blog Dashboard/Cards/DashboardJetpackSocialCardCellTests.swift:
```diff
@@ -0,0 +1,154 @@
+    func createTestBlog(isPublicizeSupported: Bool = true,
+                        hasServices: Bool = true,
+                        hasConnections: Bool = false) -> Blog {
+        let blog = NSEntityDescription.insertNewObject(forEntityName: Blog.entityName(), into: mainContext) as! Blog
+        blog.account = NSEntityDescription.insertNewObject(forEntityName: WPAccount.entityName(), into: mainContext) as? WPAccount
+        blog.capabilities = [Blog.Capability.PublishPosts.rawValue: true]
+        blog.dotComID = 12345 as NSNumber
```
Alternatively, you can take advantage of `BlogBuilder` to simplify the setup code:

```swift
var builder = BlogBuilder(mainContext)
    .withAnAccount()
    .with(blogID: 12345)

if isPublicizeSupported {
  builder = builder.with(modules: ["publicize"])
}

// ... more setup code...

return builder.build() // returns a `Blog` instance.
```

Although, since there aren't any methods yet to assign `capabilities` or `connections`, you'll need to add them to `BlogBuilder` or call `build()` early and configure them directly on the `Blog`.

wargcm replied:
That's a good call! I always forget about the builders 😆. I can add onto it to support the changes I need.

wargcm replied:
Added in https://github.com/wordpress-mobile/WordPress-iOS/pull/21180/commits/1fc9cbf27ebc0e042476f1a239d5cd4dcaf48838.

dvdchr commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/DashboardJetpackSocialCardCell.swift:
```diff
@@ -0,0 +1,231 @@
+    struct Constants {
+        static let hideNoConnectionViewKey = "dashboard-social-no-connection-view-hidden"
+        static let connectTitle = NSLocalizedString("dashboard.card.social.noconnections.title",
+                                                    value: "Connect your social profiles",
```
Based on the discussion in p1689934916528639/1689897525.127499-slack-C05362VSXFU, I suppose this should be updated to `Connect social accounts`? 

The discussion suggested that the text should be updated `Connect accounts` instead of `Connect social accounts` because the "Social" part is implied in the pre-publishing sheet (navigation title) and in the post settings (section title). Since there's no hint for "Social" in the dashboard, I think it makes sense to use `Connect social accounts` here as the card title.

```suggestion
                                                    value: "Connect social accounts",
```

wargcm replied:
`Connect social accounts` sounds like it'd be right then. I can double check with everyone just so we're all on the same page though.

wargcm replied:
Strings have been updated in https://github.com/wordpress-mobile/WordPress-iOS/pull/21180/commits/2ee12d597ce70173e1cefe1a9a53e972d7ae71e3.

dvdchr reviewed (approved):
Verified:

- The Social no connection card appears in the right place. ✅ 
- Card is hidden after adding a connection. ✅ 
- Card reappears after removing all connections. ✅ 
- Card is hidden after it's dismissed from the context menu. ✅ 
- Card reappears for a different eligible blog. ✅ 
- Tests are passing. ✅ 

Looking good! :shipit:

wargcm committed:
Use `BlogBuilder` over manually creating a `Blog`

Additionally adds the new following capabilities to
the `BlogBuilder`:

- Setting a `dotComID`
- Setting the social connections
- Setting the blog's capabilities

wargcm committed:
Update strings for the social dashboard card
